core/src/test/scala/unit/kafka/log/LogCleanerTest.scala

```diff
@@ -760,11 +725,62 @@ class LogCleanerTest extends JUnitSuite {
     assertEquals("The tombstone should be retained.", 1, log.logSegments.head.log.batches.iterator.next().lastOffset)
   }
 
+  @Test
+  def testBaseOffsetAfterCompactingV2Format(): Unit = {
```

Comment:
The heavy lifting to rewrite the log is done in the method `MemoryRecords.filterTo` which has pretty fair testing of the v2 format. Probably still not a bad idea to have an integration test here, but maybe we can try to capture some of the other stuff which is introduced in the v2 message format (e.g. producerId, producerEpoch, sequence, partitionLeaderEpoch, etc.).